kernel/base/ipc/los_rwlock.c

```diff
@@ -0,0 +1,499 @@
+{
+    RWLOCK_CB_S *pstRWLockNode;
+    UINT32   uwIndex;
+
+    LOS_ListInit(&g_stUnusedRWLockList);
+
+    if (LOSCFG_BASE_IPC_RWLOCK_LIMIT == 0)   /*lint !e506*/
+    {
+        return LOS_ERRNO_RWLOCK_MAXNUM_ZERO;
+    }
+
+    g_pstAllRWLock = (RWLOCK_CB_S *)LOS_MemAlloc(m_aucSysMem0, (LOSCFG_BASE_IPC_RWLOCK_LIMIT * sizeof(RWLOCK_CB_S)));
+    if (NULL == g_pstAllRWLock)
+    {
+        return LOS_ERRNO_RWLOCK_NO_MEMORY;
+    }
+
+    for (uwIndex = 0; uwIndex < LOSCFG_BASE_IPC_RWLOCK_LIMIT; uwIndex++)
+    {
+        pstRWLockNode                 = ((RWLOCK_CB_S *)g_pstAllRWLock) + uwIndex;
+        pstRWLockNode->ucRWLockID     = uwIndex;
+        pstRWLockNode->pstWOwner      = NULL;
+        pstRWLockNode->ucRWLockCount  = 0;
+        pstRWLockNode->ucRCount       = 0;
+        pstRWLockNode->ucRWLockStat   = OS_RWLOCK_UNUSED;
+        LOS_ListTailInsert(&g_stUnusedRWLockList, &pstRWLockNode->stRLockList);
+    }
+    return LOS_OK;
+}
+
+/*****************************************************************************
+ Function     : LOS_RWLockCreate
+ Description  : Create a rwlock,
+ Input        : None
+ Output       : puwRWLockHandle ------ RWLock operation handle
+ Return       : LOS_OK on success ,or error code on failure
+ *****************************************************************************/
+LITE_OS_SEC_TEXT_INIT  UINT32  LOS_RWLockCreate (UINT32 *puwRWLockHandle)
+{
+    UINT32      uwIntSave;
+    RWLOCK_CB_S    *pstRWLockCreated;
+    LOS_DL_LIST *pstUnusedRWLock;
+    UINT32      uwErrNo;
+    UINT32      uwErrLine;
+
+    if (NULL == puwRWLockHandle)
+    {
+        return LOS_ERRNO_RWLOCK_PTR_NULL;
+    }
+
+    uwIntSave = LOS_IntLock();
+    if (LOS_ListEmpty(&g_stUnusedRWLockList))
+    {
+        LOS_IntRestore(uwIntSave);
+        OS_GOTO_ERR_HANDLER(LOS_ERRNO_RWLOCK_ALL_BUSY);
+    }
+
+    pstUnusedRWLock                 = LOS_DL_LIST_FIRST(&(g_stUnusedRWLockList));
+    LOS_ListDelete(pstUnusedRWLock);
+    pstRWLockCreated                = (GET_RWLOCK_LIST(pstUnusedRWLock)); /*lint !e413*/
+    LOS_ListInit(&pstRWLockCreated->stRLockList);
+    LOS_ListInit(&pstRWLockCreated->stWLockList);
+    pstRWLockCreated->ucRWLockCount = 0;
+    pstRWLockCreated->ucRWLockStat  = OS_RWLOCK_USED;
+    pstRWLockCreated->ucRCount      = 0;
+    pstRWLockCreated->pstWOwner     = (LOS_TASK_CB *)NULL;
+    *puwRWLockHandle                = (UINT32)pstRWLockCreated->ucRWLockID;
+    LOS_IntRestore(uwIntSave);
+    return LOS_OK;
+ErrHandler:
+    OS_RETURN_ERROR_P2(uwErrLine, uwErrNo);
+}
+
+/*****************************************************************************
+ Function     : LOS_RWLockDelete
+ Description  : Delete a rwlock,
+ Input        : uwRWLockHandle------Mutex operation handle
+ Output       : None
+ Return       : LOS_OK on success ,or error code on failure
+ *****************************************************************************/
+LITE_OS_SEC_TEXT_INIT UINT32 LOS_RWLockDelete(UINT32 uwRWLockHandle)
```

Comment:
The interrupt lock window can be reduced.

Reply:
Can it be reduced?